flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/FlinkCalciteCatalogReader.java

```diff
@@ -62,17 +83,197 @@ public FlinkCalciteCatalogReader(
 		if (originRelOptTable == null) {
 			return null;
 		} else {
-			// Wrap FlinkTable as FlinkRelOptTable to use in query optimization.
-			FlinkTable table = originRelOptTable.unwrap(FlinkTable.class);
+			// Wrap as linkPreparingTableBase to use in query optimization.
+			CatalogSchemaTable table = originRelOptTable.unwrap(CatalogSchemaTable.class);
 			if (table != null) {
-				return FlinkRelOptTable.create(
-					originRelOptTable.getRelOptSchema(),
-					originRelOptTable.getRowType(),
+				return toPreparingTable(originRelOptTable.getRelOptSchema(),
 					originRelOptTable.getQualifiedName(),
+					originRelOptTable.getRowType(),
 					table);
 			} else {
 				return originRelOptTable;
 			}
 		}
 	}
+
+	/**
+	 * Translate this {@link CatalogSchemaTable} into Flink source table.
+	 */
+	public static FlinkPreparingTableBase toPreparingTable(RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			CatalogSchemaTable table) {
+		if (table.isTemporary()) {
+			return convertTemporaryTable(relOptSchema,
+				names,
+				rowType,
+				table.getObjectIdentifier(),
+				table.getCatalogTable(),
+				table.getStatistic(),
+				table.isStreamingMode());
+		} else {
+			return convertPermanentTable(relOptSchema,
+				names,
+				rowType,
+				table.getObjectIdentifier(),
+				table.getCatalogTable(),
+				table.getStatistic(),
+				table.getCatalog().getTableFactory().orElse(null),
+				table.isStreamingMode());
+		}
+	}
+
+	private static FlinkPreparingTableBase convertPermanentTable(
+			RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			ObjectIdentifier objectIdentifier,
+			CatalogBaseTable table,
+			FlinkStatistic statistic,
+			@Nullable TableFactory tableFactory,
+			boolean isStreamingMode) {
+		if (table instanceof QueryOperationCatalogView) {
```

Comment:
Is this possible? I was assuming the query operation can't be serialized and thus can't be permanent table.